### DISCUSSION TITLE
Timeout on Socket.Connect

### DISCUSSION THREAD

**Discussion by cincuranet** [dotnet/runtime#45426] [Q&A] [answered]
Hi. What's the recommended approach nowadays when you want to have "connect timeout" aka timeout on `Socket.Connect` method. The "old" way was to use `BeginConnect` and `AsyncWaitHandle.WaitOne` with timeout. But maybe nowadays there's a better way.

**Comment by stephentoub** [comment, marked as answer]
You can pass a CancellationToken to Socket.ConnectAsync.  That token can have cancellation requested manually or via a timer.

**Reply by cincuranet** [reply]
I don't see an overload for `Socket.ConnectAsync` with `CancellationToken` [here](https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.socket.connectasync?view=net-5.0).

**Reply by stephentoub** [reply]
All of the {Value}Task-based APIs on Socket are extension methods:
https://docs.microsoft.com/en-us/dotnet/api/system.net.sockets.sockettaskextensions.connectasync?view=net-5.0
(For .NET 6 they've been replicated as instance methods.)

**Reply by cincuranet** [reply]
Ah, great. I missed that reading just the documentation. 

But what would be an equivalent blocking/synchronous code path scenario to match the `Socket.Connect` itself? Calling `Wait` or `GetAwaiter.GetResult` might not be the smartest move here.

**Reply by stephentoub** [reply]
If the connect is synchronous, then regardless of implementation something needs to trigger something from another thread in order to interrupt the connect.  That could be using a timer to call Dispose on the Socket, for example.  Or you could just block on the ConnectAsync.

**Reply by cincuranet** [reply]
OK. Thanks.

As a library writer, I'm sweetly looking at .NET 5 and `SocketTaskExtensions.ConnectAsync`, yet returning back to multi-targetting.
